### PR TITLE
Use PV shared informer in PV controller

### PR DESCRIPTION
**Pull request by wongma7**
Use the PV shared informer, addressing (partially) https://github.com/kubernetes/kubernetes/issues/26247 . Using the PVC shared informer is not so simple because sometimes the controller wants to `Requeue` and...

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30825)

### CODE REVIEW

**Comment by wongma7** [comment]
@jsafrane can you take a look when you are not busy, after working on this I think PV can't use sharedinformer either, won't things break since the controller is updating and adding to the cache while attach/detach is reading it?

**Comment by saad-ali** [comment]
@k8s-bot ok to test

**Comment by saad-ali** [comment]
Assigning to @jsafrane

File: pkg/controller/volume/persistentvolume/controller_base.go [hunk 1]
```diff
@@ -131,32 +118,46 @@ func NewPersistentVolumeController(
...
-		if !deleted {
-			clone, err := conversion.NewCloner().DeepCopy(&volume)
-			if err != nil {
-				glog.Errorf("error cloning volume %q: %v", volume.Name, err)
-				continue
-			}
-			volumeClone := clone.(*api.PersistentVolume)
-			ctrl.storeVolumeUpdate(volumeClone)
-		}
+		ctrl.upgradeVolumeFrom1_2(pv.(*api.PersistentVolume))
```
Comment (jsafrane) [inline pkg/controller/volume/persistentvolume/controller_base.go]: please check that `pv` actually is a `PersistentVolume`:

```go
for _, obj := range volumes.List() {
  pv, ok := obj.(*api.PersistentVolume)
  if !ok { 
    ...
  } else {
    ctrl.upgradeVolumeFrom1_2
  }
```

File: pkg/controller/volume/persistentvolume/controller_base.go [hunk 2]
```diff
@@ -131,32 +118,46 @@ func NewPersistentVolumeController(
...
-		return
-	}
-	volumeList, ok := volumeListObj.(*api.PersistentVolumeList)
-	if !ok {
-		glog.Errorf("PersistentVolumeController can't initialize caches, expected list of volumes, got: %#v", volumeListObj)
-		return
-	}
-	for _, volume := range volumeList.Items {
+func (ctrl *PersistentVolumeController) initializeCaches(volumes cache.Store, claimSource cache.ListerWatcher) {
+	for _, pv := range volumes.List() {
```
Comment (jsafrane) [inline pkg/controller/volume/persistentvolume/controller_base.go]: Whole point of `initializeCaches` is to fill `ctrl.volumes` with initial content. Please restore this functionality.

There are two caches, one in the (shared) informer and second in ctrl.volumes. The controller uses the later one for all its decisions and it must be initialized here. I am quite surprised it passed unit tests.

File: pkg/controller/volume/persistentvolume/controller_base.go [hunk 3]
```diff
@@ -422,12 +423,15 @@ func (ctrl *PersistentVolumeController) deleteClaim(obj interface{}) {
 func (ctrl *PersistentVolumeController) Run() {
 	glog.V(4).Infof("starting PersistentVolumeController")
 
-	ctrl.initializeCaches(ctrl.volumeSource, ctrl.claimSource)
+	ctrl.initializeCaches(ctrl.volumes.store, ctrl.claimSource)
```
Comment (jsafrane) [inline pkg/controller/volume/persistentvolume/controller_base.go]: `initializeCaches` should copy all volumes from the shared informer into internal `ctrl.volumes`.

File: pkg/controller/volume/persistentvolume/controller_base.go [hunk 4]
```diff
@@ -63,7 +65,6 @@ func NewPersistentVolumeController(
 	}
 
 	controller := &PersistentVolumeController{
-		volumes:                       newPersistentVolumeOrderedIndex(),
```
Comment (jsafrane) [inline pkg/controller/volume/persistentvolume/controller_base.go]: this is internal controller cache, please do not remove its initialization.

File: pkg/controller/volume/persistentvolume/index.go [hunk 5]
```diff
@@ -32,10 +32,6 @@ type persistentVolumeOrderedIndex struct {
 	store cache.Indexer
 }
 
-func newPersistentVolumeOrderedIndex() persistentVolumeOrderedIndex {
```
Comment (jsafrane) [inline pkg/controller/volume/persistentvolume/index.go]: why is this moved to framework_test.go?

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 6]
```diff
@@ -413,12 +413,13 @@ func StartControllers(s *options.CMServer, kubeClient *client.Client, kubeconfig
...
 	volumeController := persistentvolumecontroller.NewPersistentVolumeController(
 		clientset.NewForConfigOrDie(restclient.AddUserAgent(kubeconfig, "persistent-volume-binder")),
+		sharedInformers.PersistentVolumes().Informer(),
 		s.PVClaimBinderSyncPeriod.Duration,
 		provisioner,
 		ProbeRecyclableVolumePlugins(s.VolumeConfiguration),
 		cloud,
 		s.ClusterName,
-		nil, nil, nil,
+		nil, nil, /* claimSource, eventRecorder */
```
Comment (jsafrane) [inline cmd/kube-controller-manager/app/controllermanager.go]: Similar changes are needed in `/contrib/mesos/pkg/controllermanager/controllermanager.go`, otherwise some unit test won't compile.

File: pkg/controller/volume/persistentvolume/controller_base.go (hunk 2 above)
Comment (wongma7) [inline pkg/controller/volume/persistentvolume/controller_base.go]: Ah I understand now :) I ripped a lot of things out, got it to pass tests, then was confused about why it was working myself.

Comment (wongma7) [inline pkg/controller/volume/persistentvolume/controller_base.go]: I think `AddEventHandler` initializes the internal cache for me by sending add events which are then handled by `addVolume`. https://github.com/kubernetes/kubernetes/blob/9a4cc9979a2650fcd047c0c493e7b362f49e384e/pkg/controller/framework/shared_informer.go#L219 So maybe I don't need to do `initializeCaches` and that's why it worked. I understand I still need to separate the internal cache and the shared one though.

**Comment by wongma7** [comment]
@jsafrane new code

**Comment by wongma7** [comment]
OK it was no disaster after all. PTAL

**Comment by jsafrane** [comment]
@k8s-bot ok to test

**Comment by jsafrane** [comment]
@k8s-bot test this issue: #31694

**Comment by jsafrane** [comment]
lgtm

**Comment by jsafrane** [comment]
@k8s-bot test this issue: #31776
